File: src/main/java/dev/obscuria/tooltips/client/tooltip/TooltipDefinition.java
package dev.obscuria.tooltips.client.tooltip;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import dev.obscuria.tooltips.client.registry.TooltipRegistries;
import dev.obscuria.tooltips.client.tooltip.filter.ItemFilter;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;

@Desugar
public record TooltipDefinition(
        int priority,
        TooltipStyle style,
        ItemFilter filter
) implements Comparable<TooltipDefinition> {
    public static TooltipDefinition fromJson(JsonObject json) {
        final int priority = JsonUtils.getInt(json, "priority");
        final TooltipStyle style = TooltipRegistries.TOOLTIP_STYLE.get(
                new ResourceLocation(JsonUtils.getString(json, "style")));
        final ItemFilter filter = ItemFilter.fromJson(JsonUtils.getJsonObject(json, "filter"));
        return new TooltipDefinition(priority, style != null ? style : TooltipStyle.EMPTY, filter);
    }

    public boolean isFor(ItemStack stack) {
        return filter.test(stack);
    }

    @Override
    public int compareTo(TooltipDefinition other) {
        return Integer.compare(priority, other.priority);
    }

    public static TooltipStyle aggregateStyleFor(ItemStack stack) {
        var style = TooltipStyle.EMPTY;
        for (var definition : TooltipRegistries.TOOLTIP_DEFINITION.listElements()) {
            if (!definition.isFor(stack)) continue;
            style = style.merge(definition.style);
        }
        return style;
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/TooltipDefinition.kt
package dev.obscuria.tooltips.client

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.tooltips.client.filter.IItemFilter
import dev.obscuria.tooltips.registry.TooltipsRegistries
import net.minecraft.world.item.ItemStack

class TooltipDefinition(
    val priority: Int,
    val style: TooltipStyle,
    val filter: IItemFilter
) : Comparable<TooltipDefinition> {

    fun isFor(stack: ItemStack): Boolean {
        return filter.test(stack)
    }

    override fun compareTo(other: TooltipDefinition): Int {
        return priority.compareTo(other.priority)
    }

    companion object {

        val DIRECT_CODEC: Codec<TooltipDefinition> = RecordCodecBuilder.create { codec ->
            codec.group(
                Codec.INT
                    .fieldOf("priority")
                    .forGetter(TooltipDefinition::priority),
                TooltipStyle.CODEC
                    .fieldOf("style")
                    .forGetter(TooltipDefinition::style),
                IItemFilter.CODEC
                    .fieldOf("filter")
                    .forGetter(TooltipDefinition::filter)
            ).apply(codec, ::TooltipDefinition)
        }

        fun aggregateStyleFor(stack: ItemStack): TooltipStyle {
            var style = TooltipStyle.EMPTY
            TooltipsRegistries.Resource.TOOLTIP_DEFINITION.listElements().forEach {
                if (it.isFor(stack)) style = style.merge(it.style)
            }
            return style
        }
    }
}